packages/terra-functional-testing/lib/express-server/express-server.js

```diff
@@ -0,0 +1,93 @@
+const fs = require('fs');
+const express = require('express');
+const Logger = require('../logger/logger');
+
+const logger = new Logger({ prefix: 'express-server' });
+
+class ExpressServer {
+  constructor(options = {}) {
+    const {
+      host,
+      index,
+      port,
+      site,
+    } = options;
+
+    this.host = host || '0.0.0.0';
+    this.index = index;
+    this.port = port || '8080';
+    this.site = site || './build';
+  }
+
+  /**
+   * Creates an express app.
+   * @returns {Object} - An instance of an express app.
+   */
+  createApp() {
+    const app = express();
+
+    app.use(express.static(this.site, {
+      ...this.index && { index: this.index },
+      extensions: ['html', 'htm'],
+    }));
+
+    app.use([/\/[^.]*$/, '/*.html?'], (_req, res, next) => {
+      // Return 404.html if provided.
+      res.status(404).sendFile('/404.html', { root: this.site }, () => {
+        // If there is an error, bail.
+        next();
+      });
+    });
+
+    return app;
+  }
+
+  /**
+   * Starts the webpack dev server.
+   * @returns {Promise} - A promise that resolves when the server has started.
+   */
+  start() {
+    // Check if the site exists prior to starting the express server.
+    if (!fs.existsSync(this.site) || (fs.lstatSync(this.site).isDirectory() && fs.readdirSync(this.site).length === 0)) {
+      logger.warn(`Cannot serve content from ${this.site} because it does not exist or it is empty.`);
+      return Promise.reject();
+    }
+
+    return new Promise((resolve, reject) => {
+      const app = this.createApp();
+
+      logger.info('Starting the express server.');
+
+      this.server = app.listen(this.port, this.host, (error) => {
+        if (error) {
+          reject(error);
```

Comment:
if we encounter an error and reject, do we want to immediately return to terminate execution?

Reply:
Can you explain this a in a little more detail? The rejection will end the lifecycle of the promise. I can wrap the resolve logic below in an `else` block to more safely guard that logic, but there's nothing to return from here.

With these changes we can return or throw a special error message to end the service execution:

https://github.com/webdriverio/webdriverio/pull/5565

Reply:
I wrapped the resolve/reject resolution in if/else blocks, I believe this will mitigate any issues:

https://github.com/cerner/terra-toolkit/pull/417/commits/732f2c36f32221c7c2ae9f20105a7a749711243f

In a future PR I'll add the new error in the appropriate places to terminate test execution if the any of the service hooks fail